runtime/vcache/nulls.go

```diff
@@ -2,42 +2,104 @@ package vcache
 
 import (
 	"io"
+	"sync"
 
-	"github.com/brimdata/zed"
-	"github.com/brimdata/zed/pkg/field"
 	"github.com/brimdata/zed/vector"
 	"github.com/brimdata/zed/vng"
+	"golang.org/x/sync/errgroup"
 )
 
-func (l *loader) loadNulls(any *vector.Any, typ zed.Type, path field.Path, m *vng.Nulls) (vector.Any, error) {
-	// The runlengths are typically small so we load them with the metadata
-	// and don't bother waiting for a reference.
-	runlens := vng.NewInt64Decoder(m.Runs, l.r) //XXX 32-bit reader?
-	var null bool
-	var off int
-	var slots []uint32
-	// In zed, nulls are generally bad and not really needed because we don't
-	// need super-wide uber schemas with lots of nulls.
-	for {
-		run, err := runlens.Next()
-		if err != nil {
-			if err == io.EOF {
-				break
-			}
-			return nil, err
+type nulls struct {
+	mu    sync.Mutex
+	meta  *vng.Nulls
+	local *vector.Bool
+	flat  *vector.Bool
+	done  bool
+}
+
+func (n *nulls) fetch(g *errgroup.Group, reader io.ReaderAt) {
+	if n == nil {
+		return
+	}
+	n.mu.Lock()
+	if n.meta == nil {
+		n.mu.Unlock()
+		return
+	}
+	n.mu.Unlock()
+	g.Go(func() error {
+		n.mu.Lock()
+		defer n.mu.Unlock()
+		if n.meta == nil {
+			return nil
 		}
-		if null {
-			for i := 0; int64(i) < run; i++ {
-				slots = append(slots, uint32(off+i))
+		length := n.meta.Count + n.meta.Values.Len()
+		n.local = vector.NewBoolEmpty(length, nil)
+		runlens := vng.NewInt64Decoder(n.meta.Runs, reader) //XXX 32-bit reader?
+		var null bool
+		var off int
+		b := n.local
+		for {
+			run, err := runlens.Next()
+			if err != nil {
+				if err == io.EOF {
+					err = nil
+				}
+				n.meta = nil
+				return err
+			}
+			if null {
+				for i := 0; int64(i) < run; i++ {
+					slot := uint32(off + i)
+					b.Set(slot)
+				}
 			}
+			off += int(run)
+			null = !null
 		}
-		off += int(run)
-		null = !null
+	})
+}
+
+func (n *nulls) flatten(parent *vector.Bool) *vector.Bool {
+	if n == nil {
+		return parent
+	}
+	n.mu.Lock()
+	defer n.mu.Unlock()
+	if n.done {
```

Comment:
Drop the done field and just check for a non-nil flat?
```suggestion
	if n.flat != nil {
```